kernel/kernel-api/src/main/java/io/delta/kernel/internal/snapshot/SnapshotManager.java

```diff
@@ -598,6 +629,22 @@ protected Optional<LogSegment> getLogSegmentForVersion(
                 );
                 throw new IllegalStateException(msg);
             }
+
+            // Reading sidecars only applies for v2 checkpoints.
+            if (newCheckpoint.format == CheckpointInstance.CheckpointFormat.V2) {
```

Comment:
Note: For a V2 checkpoint instance, do not include the compatibility file in the checkpoints in the LogSegment. Only include checkpoint manifest + sidecar files.

Reply:
This can potentially lead to incorrect results being served to the user:
1. Table A only has one checkpoint which happens to be a compatibility V2 checkpoint. Because of https://github.com/delta-io/delta/pull/2826/files#diff-12397663510cec22feeea64e8e3f0ddee114de86025e00495b02ab89cc3d2d01R69, newCheckpoint.format will be `single_checkpoint`.
2. Since we only read sidecars if format == v2, no sidecars will be read for this checkpoint and only the non-file actions will be read.
This could result in the construction of a snapshot which looks legal but is missing references to many `AddFile`s that are present in those sidecars.

Reply:
This should be fixed - now, we read the highest version checkpoint, preferring V2 over V1 checkpoints if multiple exist at the same version.